privval/signer_listener_endpoint.go

```diff
@@ -0,0 +1,294 @@
+			sl.Logger.Error("Closing Listener", "err", err)
+		}
+	}
+
+	// Stop service loop
+	sl.stopCh <- struct{}{}
+	<-sl.stoppedCh
+}
+
+// Close closes the connection
+func (sl *SignerListenerEndpoint) Close() error {
+	sl.mtx.Lock()
+	defer sl.mtx.Unlock()
+	sl.Logger.Debug("SignerListenerEndpoint: Close")
+
+	sl.dropConnection()
+	return nil
+}
+
+// IsConnected indicates if there is an active connection
+func (sl *SignerListenerEndpoint) IsConnected() bool {
+	sl.mtx.Lock()
+	defer sl.mtx.Unlock()
+	return sl.isConnected()
+}
+
+// WaitForConnection waits maxWait for a connection or returns a timeout error
+func (sl *SignerListenerEndpoint) WaitForConnection(maxWait time.Duration) error {
+	sl.mtx.Lock()
+	defer sl.mtx.Unlock()
+	return sl.ensureConnection(maxWait)
+}
+
+// SendRequest ensures there is a connection, sends a request and waits for a response
+func (sl *SignerListenerEndpoint) SendRequest(request RemoteSignerMsg) (RemoteSignerMsg, error) {
+	sl.mtx.Lock()
+	defer sl.mtx.Unlock()
+
+	sl.Logger.Debug("SignerListenerEndpoint: Send request", "connected", sl.isConnected())
+	err := sl.ensureConnection(sl.timeoutAccept)
+	if err != nil {
+		return nil, err
+	}
+
+	sl.Logger.Debug("Send request. Write")
+	err = sl.writeMessage(request)
+	if err != nil {
+		return nil, err
+	}
+
+	sl.Logger.Debug("Send request. Read")
+	res, err := sl.readMessage()
+	if err != nil {
+		sl.Logger.Debug("Read Error", "err", err)
+		return nil, err
+	}
+
+	return res, nil
+}
+
+func (sl *SignerListenerEndpoint) isConnected() bool {
+	return sl.IsRunning() && sl.conn != nil
+}
+
+func (sl *SignerListenerEndpoint) readMessage() (msg RemoteSignerMsg, err error) {
+	if !sl.isConnected() {
+		return nil, cmn.ErrorWrap(ErrListenerNoConnection, "endpoint is not connected")
+	}
+
+	// Reset read deadline
+	deadline := time.Now().Add(sl.timeoutReadWrite)
+	sl.Logger.Debug(
+		"SignerListenerEndpoint: readMessage",
+		"timeout", sl.timeoutReadWrite,
+		"deadline", deadline)
+
+	err = sl.conn.SetReadDeadline(deadline)
+	if err != nil {
+		return
+	}
+
+	const maxRemoteSignerMsgSize = 1024 * 10
+	_, err = cdc.UnmarshalBinaryLengthPrefixedReader(sl.conn, &msg, maxRemoteSignerMsgSize)
+	if _, ok := err.(timeoutError); ok {
+		err = cmn.ErrorWrap(ErrListenerTimeout, err.Error())
+		sl.dropConnection()
+	}
+
+	return
+}
+
+func (sl *SignerListenerEndpoint) writeMessage(msg RemoteSignerMsg) (err error) {
+	if !sl.isConnected() {
+		return cmn.ErrorWrap(ErrListenerNoConnection, "endpoint is not connected")
+	}
+
+	// Reset read deadline
+	deadline := time.Now().Add(sl.timeoutReadWrite)
+	sl.Logger.Debug(
+		"SignerListenerEndpoint: writeMessage",
+		"timeout", sl.timeoutReadWrite,
+		"deadline", deadline)
+
+	err = sl.conn.SetWriteDeadline(deadline)
+	if err != nil {
+		return
+	}
+
+	_, err = cdc.MarshalBinaryLengthPrefixedWriter(sl.conn, msg)
+	if _, ok := err.(timeoutError); ok {
+		err = cmn.ErrorWrap(ErrListenerTimeout, err.Error())
```

Comment:
shouldn't we sl.dropConnection() here by analogue with readMsg?

Reply:
done